[vulkan] adaptive_avg_pool2d

IvanKobzarev opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #41435 WIP - vulkan view
* #41434 [vulkan] cat op (concatenate)
* #41380 [vulkan] inplace add_, relu_
* #41379 [vulkan] max_pool2d
* #41224 [vulkan][asan] Fix Invalid Memory ops
* #41223 [vulkan] reshape op
* #41222 [vulkan] support add for dim < 4
* #41221 [vulkan] mm op through addmm
* **#41220 [vulkan] adaptive_avg_pool2d**

Differential Revision: [D22754943](https://our.internmc.facebook.com/intern/diff/D22754943)

IvanKobzarev committed:
[vulkan] adaptive_avg_pool2d

[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] adaptive_avg_pool2d"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] adaptive_avg_pool2d"




[ghstack-poisoned]

AshkanAliabadi reviewed (approved):
Please consider using const anywhere possible.  I think it is one of the few tools that can help us make a language as unsafe as C and C++ a little bit safer.  Also even if there is a 0.1% possibility that use of const can somehow result in generation of a more optimized code somewhere, well I would take that too. :)

IvanKobzarev commented on aten/src/ATen/native/vulkan/glsl/adaptive_avg_pool2d.glsl:
```diff
@@ -0,0 +1,41 @@
+#version 450 core
+layout(std430) buffer;
+layout(std430) uniform;
+layout(set = 0, rgba16f, binding = 0) writeonly mediump uniform image3D uOutput;
```
Yes, I experimented with highp and mediump. I think for now we should use highp everywhere and introduce parametrization that we can use mediump. I will change it to highp. Will also need to change it in other glsl files in separate PR.

IvanKobzarev committed:
Update on "[vulkan] adaptive_avg_pool2d"




[ghstack-poisoned]

IvanKobzarev committed:
Update on "[vulkan] adaptive_avg_pool2d"


Differential Revision: [D22754943](https://our.internmc.facebook.com/intern/diff/D22754943)

[ghstack-poisoned]
